src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorProjectChangePublisher.cs

```diff
@@ -267,12 +274,29 @@ protected virtual bool ShouldSerialize(string configurationFilePath)
 
             var status = _operationProgressStatusService?.GetStageStatusForSolutionLoad(CommonOperationProgressStageIds.Intellisense);
 
+            // Don't serialize our understanding until we've parsed at least one document,
+            // otherwise our understanding is incomplete
+            if (!_documentsProcessed)
+            {
+                foreach (var document in projectSnapshot.DocumentFilePaths)
+                {
+                    var fileName = Path.GetFileNameWithoutExtension(document);
+
+                    if (projectSnapshot.TagHelpers.Any(t => t.Name.EndsWith("." + fileName, StringComparison.OrdinalIgnoreCase)))
```

Comment:
Am I reading this right? If there are no components in the project then it will never serialize?

Reply:
True, but if there are no components in your project you're not going to have any Razor tooling anyway right?

Reply:
I guess that makes sense. In my mind a "component" is different from a "page", but I'm guessing thats just my lack of Razor understanding.

Reply:
When restricting check to non-Blazor projects you can look at document file kinds like:

```C#
var documentSnapshot = projectSnapshot.GetDocument(document);
if (documentSnapshot.FileKind == AspNetCore.Razor.Language.FileKinds.Component)
```

Reply:
You're not wrong @davidwengier, for non-Blazor projects  we'll need to make sure to override this check. Reason being a Foo.cshtml will never be represented in the TagHelper list by default. TagHelpers operated a bit differently in the past 😄